File: src/main/java/oscar/cp/constraints/TableData.java
/*******************************************************************************
 * This file is part of OscaR (Scala in OR).
 *  
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 * 
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/gpl-3.0.html
 ******************************************************************************/

package oscar.cp.constraints;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * Table Data used to build a table constraint and specify the acceptable tuples
 * @author pschaus@gmail.com
 */
public class TableData {

	int arity;

	ArrayList<Integer> [] data;

	int [][] firstSupport;
	int [][] nextSupport;
	int [] min;
	int [] max;
	
	int [][] tuples = null;

	/**
	 *
	 * @param arity of the tupples than can be added to this table
	 */
	@SuppressWarnings("unchecked")
	public TableData(int arity){
		this.arity = arity;

		data = new ArrayList[arity];
		for (int i = 0; i < data.length; i++) {
			data[i] = new ArrayList<Integer>();
		}

		firstSupport = new int [arity] [];
		nextSupport = new int [arity] [];
		min = new int[arity];
		max = new int [arity];


	}
	
	/**
	 * @return one array entry per tuple
	 */
	public int [][] getTuples() {
		if (tuples != null) return tuples;
		int [][] res = new int[data[0].size()][arity];
		for (int i = 0; i < res.length; i++) {
			for (int a = 0; a < arity; a++) {
				res[i][a] = data[a].get(i);
			}
		}
		tuples = res;
		return tuples;
	}

	protected int getMin(int i){
		return min[i];
	}

	protected int getMax(int i){
		return max[i];
	}

	/**
	 * Adds a possible tuple
	 * @param tuple must have a length arity
	 */	
	public void addTuple(int ...tuple){
		assert(arity == tuple.length);
		for (int i = 0; i < tuple.length; i++) {
			data[i].add(tuple[i]);
		}	
	}

	protected int getValue(int tuple, int index){
		return data[index].get(tuple);
	}

	protected void printTuple(int tuple){
		for (int i = 0; i < arity; i++) {
			System.out.print(data[i].get(tuple)+" ");
		} System.out.println();
	}

	protected void setUp(){
		for (int i = 0; i < arity; i++) {
			buildSupport(i);
		}
	}

	protected void buildSupport(int i){
		int mini = Integer.MAX_VALUE;
		int maxi = Integer.MIN_VALUE;
		for(int j = 0; j < data[i].size(); j++) {
			if (data[i].get(j) < mini) {
				mini = data[i].get(j);
			}
			if (data[i].get(j) > maxi) {
				maxi = data[i].get(j);
			}
		}
		min[i] = mini;
		max[i] = maxi;

		firstSupport[i] = new int [maxi-mini+1];
		for (int j = 0; j < firstSupport[i].length; j++) {
			firstSupport[i][j] = -1;
		}
		nextSupport[i] = new int[data[i].size()];
		for (int j = 0; j < nextSupport[i].length; j++) {
			nextSupport[i][j] = -1;
		}

		for (int j = 0; j < data[i].size(); j++) {
			int v = data[i].get(j);
			nextSupport[i][j] = firstSupport[i][v-min[i]];
			firstSupport[i][v-min[i]] = j;
		}	
	}

	protected int getFirstSupport(int i,int value){
		return firstSupport[i][value-min[i]];
	}

	protected boolean hasFirstSupport(int i,int value){
		return firstSupport[i][value-min[i]] != -1;
	}

	protected boolean hasNextSupport(int i,int tuple){
		if (tuple < 0) {
			return false;
		}
		return nextSupport[i][tuple] != -1;
	}

	protected int getNextSupport(int i,int tuple){
		if (tuple < 0) {
			return tuple;
		}
		return nextSupport[i][tuple];
	}
	
	@Override
	public String toString() {
		String res = "";
		for (int i = 0; i < data[0].size(); i++) {
			res += "( ";
			for (ArrayList<Integer> col : data) {
				res += col.get(i)+ " ";
			}
			res += ") , ";
		}
		return res;	
	}


}

File: src/main/scala/oscar/cp/constraints/TableData.scala
/*******************************************************************************
 * This file is part of OscaR (Scala in OR).
 *  
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 * 
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/gpl-3.0.html
 ******************************************************************************/

package oscar.cp.constraints

import scala.collection.mutable.ArrayBuffer

class TableData(val arity: Int) {
  
  /*
  def this[T <: Product](items: Iterable[T]) {
    this(items.head.arity)
    items.foreach(t => add(t.iterator:_*))
  }
  */
 
  
  val data = Array.fill(arity)(ArrayBuffer[Int]())
  
  val firstSup = Array.fill(arity)(Array[Int]())
  
  val nextSup = Array.fill(arity)(Array[Int]())
  
  val minVal = Array.fill(arity)(0)
  val maxVal = Array.fill(arity)(0)
  
  def min(i: Int): Int = minVal(i)
  def max(i: Int): Int = maxVal(i)
  
 /**
   * Adds a possible tuple
   * @param tuple must have a length arity
   */	
  def add(tuple: Int*) {
		assert(arity == tuple.length,{println("wrong arrity:"+tuple.length)})
		tuple.zipWithIndex.foreach{case(v,i) => data(i) + tuple(i)} 
  }
  
  /**
   * retrieve entry i of tuple t
   */
  def apply(t: Int, i: Int) = {
    data(i)(t) 
  }
  
  def setup() = {
    for (i <- 0 until arity) {
      minVal(i) = data(i).min
      maxVal(i) = data(i).max
      val nVal = maxVal(i)- minVal(i) + 1
      firstSup(i) = Array.fill(nVal) (-1)
      nextSup(i) = Array.fill(data(i).size)(-1)
      for ((v,j) <- data(i).zipWithIndex) {
	  	nextSup(i)(j) = firstSup(i)(v-min(i))
	  	firstSup(i)(v-min(i)) = j
	  } 
    }
  }
  
  def firstSupport(i: Int, value: Int) = firstSup(i)(value-min(i))
  
  def hasFirstSupport(i: Int, value: Int) = firstSupport(i,value) != -1
  
  def nextSupport(i: Int,t: Int) = {
		if (t < 0) {
			t
		} else {
		  nextSup(i)(t)
		}
  }
  
  def hasNextSupport(i: Int, t: Int) = {
	  if (t < 0) false
	  else nextSup(i)(t) != -1
  }
}
